src/audio/codec_adapter/codec/dts.c

```diff
@@ -0,0 +1,365 @@
+	unsigned int alignment)
+{
+	struct comp_dev *dev = (struct comp_dev *)dev_void;
+	void *pMem;
+
+	comp_dbg(dev, "dts_effect_allocate_codec_memory() start");
+
+	pMem = codec_allocate_memory(dev, (uint32_t)length, (uint32_t)alignment);
+
+	if (pMem == NULL)
+		comp_err(dev,
+			"dts_effect_allocate_codec_memory() failed to allocate %d bytes", length);
+
+	comp_dbg(dev, "dts_effect_allocate_codec_memory() done");
+	return pMem;
+}
+
+static int dts_effect_convert_sof_interface_result(struct comp_dev *dev,
+	DtsSofInterfaceResult dts_result)
+{
+	int ret;
+
+	switch (dts_result) {
+	case DTS_SOF_INTERFACE_RESULT_SUCCESS:
+		ret = 0;
+		break;
+	case DTS_SOF_INTERFACE_RESULT_ERROR_NO_MEMORY:
+		ret = -ENOMEM;
+		break;
+	case DTS_SOF_INTERFACE_RESULT_ERROR_DTS_INTERNAL_MODULE_ERROR:
+		ret = -EIO;
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int dts_effect_populate_buffer_configuration(struct comp_dev *dev,
+	DtsSofInterfaceBufferConfiguration *buffer_config)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	const struct audio_stream *stream;
+	DtsSofInterfaceBufferLayout buffer_layout;
+	DtsSofInterfaceBufferFormat buffer_format;
+
+	comp_dbg(dev, "dts_effect_populate_buffer_configuration() start");
+
+	if (cd->ca_source == NULL)
+		return -EINVAL;
+
+	stream = &cd->ca_source->stream;
+
+	switch (cd->ca_source->buffer_fmt) {
+	case SOF_IPC_BUFFER_INTERLEAVED:
+		buffer_layout = DTS_SOF_INTERFACE_BUFFER_LAYOUT_INTERLEAVED;
+		break;
+	case SOF_IPC_BUFFER_NONINTERLEAVED:
+		buffer_layout = DTS_SOF_INTERFACE_BUFFER_LAYOUT_NONINTERLEAVED;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		buffer_format = DTS_SOF_INTERFACE_BUFFER_FORMAT_SINT16LE;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		buffer_format = DTS_SOF_INTERFACE_BUFFER_FORMAT_SINT24LE;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		buffer_format = DTS_SOF_INTERFACE_BUFFER_FORMAT_SINT32LE;
+		break;
+	case SOF_IPC_FRAME_FLOAT:
+		buffer_format = DTS_SOF_INTERFACE_BUFFER_FORMAT_FLOAT32;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	buffer_config->bufferLayout = buffer_layout;
+	buffer_config->bufferFormat = buffer_format;
+	buffer_config->sampleRate = stream->rate;
+	buffer_config->numChannels = stream->channels;
+	buffer_config->totalBufferLengthInBytes = cd->period_bytes;
+
+	comp_dbg(dev, "dts_effect_populate_buffer_configuration() done");
+
+	return 0;
+}
+
+int dts_codec_init(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	DtsSofInterfaceResult dts_result;
+	DtsSofInterfaceVersionInfo interface_version;
+	DtsSofInterfaceVersionInfo sdk_version;
+
+	comp_dbg(dev, "dts_codec_init() start");
+
+	dts_result = dtsSofInterfaceInit((DtsSofInterfaceInst **)&(codec->private),
+		dts_effect_allocate_codec_memory, dev);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	if (!ret) {
```

Comment:
nit, flip this with case below and return on error to reduce indent

Reply:
Good point. Now removed.